### ISSUE TITLE
Opus files' metadata isn't being captured by the bot

### ISSUE THREAD

**Issue by felix91gr** [azlux/botamusique#181]
**Describe the bug**

`.opus` files' metadata is not being read by the bot.

**Affected version**

Current `master` branch, although I've seen (but not tracked thoroughly as I have done with `master`) this bug in current stable versions as well.

Commit: `e981782`

**To Reproduce**
1. Encode a file in `opus`
2. Add metadata to it. Title and Artist are good enough for demostration.
3. Add the song to your Music Library
4. Reload your Library
5. Look for the new song.
6. You should see none of the metadata you added.

**Expected behavior**

For it to behave just like the `mp3` files, that is, to show the metadata in the Library and Playlists.

**Screenshots**

This is the original `opus` file in my Library:

![image](https://user-images.githubusercontent.com/11747623/86694465-a1b4a380-bfd9-11ea-840f-519a577f41f4.png)

And this is how it looked after running `ffmpeg` to convert it to `mp3`:

![image](https://user-images.githubusercontent.com/11747623/86694577-be50db80-bfd9-11ea-8911-17bedd8ef287.png)

This is how I converted from the song `opus` to `mp3`:

`ffmpeg -i Test\ 📚\ \(ChilledCow\).opus  -vn -ar 44100 -ac 2 -b:a 192k -map_metadata 0:s:a:0 ffmpeg_test_mapped.mp3`

**Comment by felix91gr** [comment]
Also, this might be a FAQ:

> Why are you not storing your library in `mp3`?

To this, the answer is very simple!

> I prefer to have it in `opus` since it will be transcoded to `opus` anyways in order to send it through Mumble. I prefer to pre-encode it myself, because then I can ensure best quality in the audio stream, and also less resource usage from the unnecessary transcoding :)

**Comment by TerryGeng** [comment]
This is actually not a bug but a feature request. Currently, the only formats supported for auto-metadata loading are mp3 and m4a(mp4).

The library we use to read metadata is mutagen. It supports other formats but the interface is subtly different. I think you can have a look at https://github.com/azlux/botamusique/blob/e981782dd77624501242c815f49c26ebea77f2d6/media/file.py#L86 and you are welcomed to make a PR. I can also look into this feature but it may take some time.

**Comment by felix91gr** [comment]
Thank you for taking a look at this Terry :heart: 

I didn't know the auto-metadata thing. Do you think the quality penalty for transcoding mp3 would be too high? Because if this is intended, I could use mp3... but I don't know if it will make the listening experience worse.

I was also concerned about performance, but I guess transcoding mp3 into opus isn't that expensive?

So anyway... I have this almost ready for a PR. `mutagen` is fairly easy to use. But I'm having trouble with the Opus cover art in my songs... there seems to be some padding in the `base64` encoding that interferes with the image decoding when anything tries to render it. It's so strange. 

This is what I have, anyway:

```python
elif ext == "opus":
    tags = mutagen.File(self.uri())
    # This works
    if 'title' in tags:
        self.title = tags['title'][0]
    # This works also
    if 'artist' in tags:
        self.artist = tags['artist'][0]

    if im is None:
        # This doesn't really work yet.
        if 'metadata_block_picture' in tags:
            pic_as_base64 = tags['metadata_block_picture'][0]
            as_binary = base64.b64decode(pic_as_base64)
            im = Image.open(BytesIO(as_binary))
```

**Comment by felix91gr** [comment]
Like, for some reason, this works:

```python
    if im is None:
        # This doesn't really work yet.
        if 'metadata_block_picture' in tags:
            pic_as_base64 = tags['metadata_block_picture'][0]
            as_binary = base64.b64decode(pic_as_base64)
            im = Image.open(BytesIO(as_binary))
```

The embedding of images in `opus` is weird. I'm not 100% sure that this isn't my music tagging program's fault, but I'll check on the RFC's of `opus` and see if there's any mention of this weird header/padding thing that is not an image.

**Comment by felix91gr** [comment]
Sorry, pasted it wrong. Now it's taking the correct slice.

**Comment by felix91gr** [comment]
Okay, I think I've got it. This is the code that we want:

```python
elif ext == "opus":
    tags = mutagen.File(self.uri())
    if 'title' in tags:
        self.title = tags['title'][0]
    if 'artist' in tags:
        self.artist = tags['artist'][0]

    if im is None:
        if 'metadata_block_picture' in tags:
            pic_as_base64 = tags['metadata_block_picture'][0]
            as_flac_picture = mutagen.flac.Picture(base64.b64decode(pic_as_base64))
            im = Image.open(BytesIO(as_flac_picture.data))
```

**Comment by felix91gr** [comment]
Turns out, the image isn't encoded entirely in `base64`. It's a binary encoding based on `base64`... which has a header, and then an image. Weird stuff but hey, `mutagen` saves the day :3

**Comment by felix91gr** [comment]
Okay, TODO for me @felix91gr:

1. Make a PR with this
2. Check if this works for Ogg Vorbis files as well
3. Check if this works for FLAC files as well
4. Wait. Ask if FLAC is supported in Botamusique first, maybe it's not necessary :P

**Comment by TerryGeng** [comment]
As long as ffmpeg can decode a file into PCM, then that format is supported :)

**Comment by felix91gr** [comment]
Holy. Crap. 🤩

**Comment by felix91gr** [comment]
Closed by #184
